README.md
# sexp-parser
Use Lisp s-expressions as a data format like JSON. Come on, you know you wanna.

This is an alternative to using JSON and works for many of the same usecases i.e. encoding data structures in strings for storage, transmission, or inspection.

This is what happens when you make a Lisp enthusiast write javascript all day.

Example comparison between JSON and s-expressions
```
{"number":123,"digits":["1","2","3"],"attributes":{"even":false,"positive":true}}
```
```
(:number 123 :digits ("1" "2" "3") :attributes (:even NIL :positive T))
```
```
{
  number: 123,
  digits: ['1', '2', '3'],
  attributes: {
    even: false,
    positive: true
  }
}
```
```
(:number 123
 :digits ("1" "2" "3")
 :attributes (:even NIL
              :positive T))
```

Technically, I'm implementing a subset of s-expressions, with specific semantics around key-value type data (a.k.a hashes, dictionaries, maps, named-tuples, etc.) They must be of the format:
```
(:key-1 value-1 :key-n value-n)
```
A list that does not match these semantics will be interpreted as an array (list, vector, sequence, etc.).

If an implementation language does not support symbols, the symbol will be encoded as a string, including a leading colon.
If an implementation language does not support hyphens in symbol names, they will be transcoded to underscores. (It might be nice to make this configurable in the future, as some languages "sorta" support them. Javascript can have hyphens in object keys as long as they are strings but can't have hyphens in variables, and I don't really want to cause problems with that down the line.)

## Why?

For fun, of course. When I just started programming, all data between servers and clients was XML. Then JSON was adopted widely for data transfer. It was more readable, more writable, and just felt right.

I think we can go one step further. Lisp has a very minimal syntax, and in some cases can even represent the same data in fewer bytes. JSON still frustrates me from time to time with its super strict syntax. Since I've been using Lisp a bit more, I thought this would be a cool way to bring some Lisp in to a world of javascript objects and php arrays.

Plus, I love playing with bunches of programming languages, and this gives me a chance to do just that!

## Supported Languages:
- Javascript
- Crystal
- PHP
- Clojure
- (Maybe another Lisp? CL, Scheme, Racket, etc?)
- more to come!


clojure/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Added
Added a clojure version of sexp-parser


[Unreleased]: https://github.com/your-name/sexp-parser/compare/0.1.1...HEAD


javascript/index_test.js

/********************************************************************/
/// Encoder code finish!
/// Testing time!!!
/********************************************************************/

const {
    chompWhitespace,
    findEndOfSymbol,

    tryParseBoolean,
    tryParseString,
    tryParseSymbol,
    tryParseNumber,
    tryParseSexp,
    tryParseObject,

    isParensBalanced,
    isQuotesBalanced,

    parse: parseExpression,

    encodeAtom,
    encodeArray,
    encode
} = require( './index.js' );

const test = require( 'tape' );

//TODO try parse bool!!!!
test( 'tryParseBoolean', ( t ) => {
    t.equal( tryParseBoolean( 'T', 0 ).data, true );

    t.equal( tryParseBoolean( 'NIL', 0 ).data, false );

    t.end();
} );

test( 'tryParseString', ( t ) => {
    t.equal( tryParseString( '"asdf"', 0 ).data, 'asdf',
             'parse string "asdf" data' );
    t.equal( tryParseString( '"asdf"', 0 ).next, 6,
             'parse string "asdf" next' );

    t.equal( tryParseString( '"asdf")', 0 ).data, 'asdf',
             'parse string "asdf)" data' );
    t.equal( tryParseString( '"asdf")', 0 ).next, 6,
             'parse string "asdf)" next' );

    t.equal( tryParseString( '\'asdf\'', 0 ).data, null,
             'string with single quotes' );

    t.equal( tryParseString( '"\""', 0 ).data, '"',
           'an escaped quote' );

    t.equal( tryParseString( '"\\"', 0 ).data, '\\',
           'single literal backslash' );
    t.end();
} );

test( 'tryParseSymbol', ( t ) => {
    t.equal( tryParseSymbol( ':foo', 0 ).data, ':foo',
             'symbol data' );
    t.equal( tryParseSymbol( ':foo', 0 ).next, 4,
             'symbol length' );

    t.equal( tryParseSymbol( ':bar)', 0 ).data, ':bar',
             'symbol data trailing paren' );
    t.equal( tryParseSymbol( ':bar)', 0 ).next, 4,
             'symbol next trailing paren' );

    t.equal( tryParseSymbol( ':cAsEsEnSiTiVe', 0 ).data, ':cAsEsEnSiTiVe',
             'should preserve case' );

    t.equal( tryParseSymbol( 'asdf', 0 ).data, null,
             'symbol without colon' );
    t.end();
} );

test( 'tryParseNumber', ( t ) => {
    t.equal( tryParseNumber( '12', 0 ).data, 12,
             'data 12' );
    t.equal( tryParseNumber( '12', 0 ).next, 2,
             'length 12 is 2' );

    t.equal( tryParseNumber( '-1', 0 ).data, -1,
             'data -1' );
    t.equal( tryParseNumber( '-1', 0 ).next, 2,
             'length -1 is 2' );

    t.equal( tryParseNumber( '0.5', 0 ).data, 0.5,
             'data 0.5' );
    t.equal( tryParseNumber( '0.5', 0 ).next, 3,
             'length 0.5 is 3' );

    t.equal( tryParseNumber( '.5', 0 ).data, 0.5,
             'data .5' );
    t.equal( tryParseNumber( '.5', 0 ).next, 2,
             'length .5 is 2' );

    t.equal( tryParseNumber( '-0.5', 0 ).data, -0.5,
             'data -0.5' );
    t.equal( tryParseNumber( '-0.5', 0 ).next, 4,
             'length -0.5 is 4' );

    t.equal( tryParseNumber( 'foo 12 bar', 4 ).data, 12,
             'data "foo 12 bar" is 12' );
    t.equal( tryParseNumber( 'foo 12 bar', 4 ).next, 6,
             'length "foo 12 bar" is 6' );

    t.equal( tryParseNumber( '2)', 0 ).data, 2,
             'trailing paren' );
    t.equal( tryParseNumber( '2)', 0 ).next, 1,
             'trailing paren length' );

    t.end();
} );

test( 'tryParseSexp', ( t ) => {
    t.deepEqual( tryParseSexp( '(1 2 3)', 0 ).data, [1, 2, 3] );
    t.equal( tryParseSexp( '(1 2 3)', 0 ).next, 7 );

    t.deepEqual( tryParseSexp( '()', 0).data, [] );
    t.deepEqual( tryParseSexp( '()', 0).next, 2 );

    t.test( 'nested expressions', ( st ) => {
        st.deepEqual( tryParseSexp( '(1 (2 3) 4)', 0 ).data, [1, [2, 3], 4] );
        st.deepEqual( tryParseSexp( '(1 (2 3) 4)', 0 ).next, 11 );

        st.deepEqual( tryParseSexp( '(())', 0 ).data, [[]] );
        st.deepEqual( tryParseSexp( '(())', 0 ).next, 4 );

        st.deepEqual( tryParseSexp( '( ( ) )', 0 ).data, [[]] );
        st.deepEqual( tryParseSexp( '( ( ) )', 0 ).next, 7 );

        st.deepEqual( tryParseSexp( '(() () ())', 0 ).data, [[], [], []],
                      '(() () ())' );

        st.deepEqual( tryParseSexp( '( ( 2 ) 3)', 0).data, [[2], 3] );

        st.end();
    } );

    t.test( 'parsing key-value objects', ( st ) => {
        let input, expected;

        input = '(:a 1)';
        expected = {a: 1};
        st.deepEqual( tryParseSexp( input, 0 ).data, expected );

        input = '(:foo 123 :bar "asd")';
        expected = { foo: 123, bar: 'asd' };
        st.deepEqual( tryParseSexp( input, 0 ).data, expected );

        input = '(:a (:b (:c 69)))';
        expected = { a: { b: { c: 69 } } };
        st.deepEqual( tryParseSexp( input, 0 ).data, expected,
                      'parsing nested objects: (:a (:b (:c 69)))' );

        st.deepEqual( tryParseSexp( '(:key "value")', 0 ).data, {key: 'value'},
                      'Parsing something object-like.' );


        st.end();
    } );

    t.end();
} );

test( 'tryParseObject', ( t ) => {
    let array = [':foo', 123, ':bar', 'asd'];
    let expected = { foo: 123, bar: 'asd' };
    t.deepEqual( tryParseObject( array ), expected );

    array = [ ':A', [ ':B', [ ':C', 69 ] ] ];
    expected = { A: [ ':B', [ ':C', 69 ] ] };
    t.deepEqual( tryParseObject( array ), expected,
                 'should only do the first level of an array' );

    t.end();
} );

test( 'chompWhitespace', ( t ) => {
    t.equal( chompWhitespace( ' a', 0 ), 1,
             '" a" should give position 1' );

    t.end();
} );

test( 'findEndOfSymbol', ( t ) => {
    t.equal( findEndOfSymbol( 'a ', 0 ), 1 );

    t.end();
} );

test( 'isParensBalanced', ( t ) => {
    t.ok( isParensBalanced( '()' ), '()' );
    t.ok( isParensBalanced( '(()()(()))' ), '(()()(()))' );
    t.ok( isParensBalanced( '(works (with ( stuff )in )between )' ), '(works (with ( stuff )in )between )' );

    t.notOk( isParensBalanced( ')(' ), 'close before open' );
    t.notOk( isParensBalanced( '(' ), 'just one open' );
    t.notOk( isParensBalanced( ')' ), 'just one close' );
    t.notOk( isParensBalanced( '())' ), '())' );
    t.notOk( isParensBalanced( '(()' ), '(()' );

    t.end();
} );

test( 'isQuotesBalanced', ( t ) => {
    t.ok( isQuotesBalanced( '' ), 'zero quotes' );
    t.ok( isQuotesBalanced( '""' ), 'two quotes' );

    t.notOk( isQuotesBalanced( '"""' ), 'three quotes' );

    t.end();
} );

test( 'encodeAtom', ( t ) => {
    t.equal( encodeAtom( "asd" ), '"asd"' );
    t.equal( encodeAtom( 69 ), '69' );

    t.equal( encodeAtom( true ), 'T', 'true should be T' );
    t.equal( encodeAtom( false ), 'NIL', 'false should be NIL' );

    t.end();
} );

test( 'encodeArray', ( t ) => {
    t.equal( encodeArray( [] ), '()' );
    t.equal( encodeArray( [1, 2, 3] ), '(1 2 3)' );

    t.equal( encodeArray( [[1, 2], [3, 4]] ), '((1 2) (3 4))' );

    t.end();
} );

test( 'encode', ( t ) => {
    let actual = encode( {
        number: 123,
        digits: ['1', '2', '3'],
        attributes: {
            even: false,
            positive: true
        }
    } );
    let expected = '(:number 123 :digits ("1" "2" "3") :attributes (:even NIL :positive T))';

    t.equal( actual, expected, 'large encoding sample' );

    // t.equal( 'encode',
    t.end();
} );

test( 'large round trip test', ( t ) => {
    const largeDataPayload = [
        {
            "_id": "5c77256b120e71539a4543a7",
            "index": 0,
            "guid": "c9815eb2-fdc5-42f8-bbb3-c2dadaf4c1b8",
            "isActive": true,
            "balance": "$1,556.23",
            "picture": "http://placehold.it/32x32",
            "age": 40,
            "eyeColor": "brown",
            "name": "Dora Weiss",
            "gender": "female",
            "company": "CIRCUM",
            "email": "doraweiss@circum.com",
            "phone": "+1 (957) 431-3727",
            "address": "219 Chase Court, Hebron, New Jersey, 6091",
            "about": "Quis ipsum eu deserunt ut mollit est eu duis commodo reprehenderit cupidatat nulla ad aliquip. Tempor nulla in consequat est voluptate veniam excepteur pariatur sunt sint. Veniam laborum aliqua esse elit qui est exercitation velit. Mollit qui exercitation nulla elit ex ad aute laborum reprehenderit. Quis ipsum proident velit do. Dolor dolore incididunt aliquip culpa cillum enim. Id incididunt ullamco aliquip id mollit elit ullamco qui duis ut mollit.\r\n",
            "registered": "2014-08-01T12:47:41 +07:00",
            "latitude": -82.633003,
            "longitude": 4.637068,
            "tags": [
                "fugiat",
                "eu",
                "Lorem",
                "laborum",
                "fugiat",
                "pariatur",
                "ex"
            ],
            "friends": [
                {
                    "id": 0,
                    "name": "Lynette Mccarthy"
                },
                {
                    "id": 1,
                    "name": "Leila Medina"
                },
                {
                    "id": 2,
                    "name": "Kasey Moss"
                }
            ],
            "greeting": "Hello, Dora Weiss! You have 9 unread messages.",
            "favoriteFruit": "banana"
        },
        {
            "_id": "5c77256ba6e9551c135fc72f",
            "index": 1,
            "guid": "ec2b621f-e35a-4447-9ecb-c3bee2feeeea",
            "isActive": true,
            "balance": "$1,782.51",
            "picture": "http://placehold.it/32x32",
            "age": 30,
            "eyeColor": "blue",
            "name": "Karen Hernandez",
            "gender": "female",
            "company": "CONCILITY",
            "email": "karenhernandez@concility.com",
            "phone": "+1 (853) 530-3600",
            "address": "478 Schenck Street, Roland, Federated States Of Micronesia, 8478",
            "about": "Esse id do irure ipsum quis occaecat cillum pariatur est ipsum esse exercitation reprehenderit aliqua. Ea nostrud fugiat sunt sunt. Ut aute voluptate consequat occaecat cupidatat magna irure. Occaecat sunt fugiat laboris est labore laborum dolor reprehenderit irure ipsum officia voluptate id nisi. Duis consequat minim nostrud magna nostrud consequat eiusmod nulla qui sunt incididunt cupidatat. Lorem ullamco aliqua officia adipisicing eiusmod labore incididunt irure aute nulla aliquip laborum cillum.\r\n",
            "registered": "2016-03-14T11:34:25 +07:00",
            "latitude": -5.325438,
            "longitude": 18.0786,
            "tags": [
                "exercitation",
                "sit",
                "ad",
                "sit",
                "fugiat",
                "enim",
                "non"
            ],
            "friends": [
                {
                    "id": 0,
                    "name": "Marilyn Mooney"
                },
                {
                    "id": 1,
                    "name": "Cindy Nunez"
                },
                {
                    "id": 2,
                    "name": "Ramirez Gilmore"
                }
            ],
            "greeting": "Hello, Karen Hernandez! You have 4 unread messages.",
            "favoriteFruit": "strawberry"
        },
        {
            "_id": "5c77256bef0f2c88b63e1c2a",
            "index": 2,
            "guid": "7f91e154-37c3-40ca-b240-71854027eb7f",
            "isActive": true,
            "balance": "$2,210.88",
            "picture": "http://placehold.it/32x32",
            "age": 26,
            "eyeColor": "brown",
            "name": "Short Page",
            "gender": "male",
            "company": "PLEXIA",
            "email": "shortpage@plexia.com",
            "phone": "+1 (834) 421-2685",
            "address": "129 Taaffe Place, Belmont, New Mexico, 7156",
            "about": "Ipsum minim eiusmod cupidatat irure. Adipisicing voluptate ex mollit reprehenderit. Occaecat aliqua proident enim cillum adipisicing. Do adipisicing incididunt sit non cillum. Aute duis dolor incididunt nulla sit duis cillum non id.\r\n",
            "registered": "2019-01-20T12:58:27 +08:00",
            "latitude": -10.195242,
            "longitude": 7.95824,
            "tags": [
                "occaecat",
                "proident",
                "pariatur",
                "officia",
                "eiusmod",
                "velit",
                "laboris"
            ],
            "friends": [
                {
                    "id": 0,
                    "name": "Naomi Jennings"
                },
                {
                    "id": 1,
                    "name": "Blankenship Pruitt"
                },
                {
                    "id": 2,
                    "name": "Collins Carney"
                }
            ],
            "greeting": "Hello, Short Page! You have 3 unread messages.",
            "favoriteFruit": "strawberry"
        },
        {
            "_id": "5c77256b4b4fbbd1e17ea23e",
            "index": 3,
            "guid": "39955cf0-3a93-4fa8-a85f-882bdcfc43c7",
            "isActive": true,
            "balance": "$3,613.59",
            "picture": "http://placehold.it/32x32",
            "age": 30,
            "eyeColor": "green",
            "name": "Katelyn Bruce",
            "gender": "female",
            "company": "ZOINAGE",
            "email": "katelynbruce@zoinage.com",
            "phone": "+1 (962) 514-3834",
            "address": "201 Irwin Street, Succasunna, Texas, 1045",
            "about": "Commodo elit laborum amet mollit. Non aute irure esse esse est anim ea mollit qui commodo. Do dolor occaecat Lorem incididunt qui anim commodo sunt. Occaecat cupidatat occaecat eiusmod dolore officia excepteur culpa laborum id eiusmod velit. Voluptate non cillum id nulla deserunt consectetur esse sunt quis. Eiusmod et anim occaecat sunt aute do aliqua id reprehenderit voluptate occaecat duis aliqua. Veniam nostrud velit laboris excepteur occaecat est elit.\r\n",
            "registered": "2014-08-27T03:12:23 +07:00",
            "latitude": 61.029178,
            "longitude": 51.670902,
            "tags": [
                "consectetur",
                "deserunt",
                "irure",
                "occaecat",
                "voluptate",
                "sint",
                "esse"
            ],
            "friends": [
                {
                    "id": 0,
                    "name": "Daniel Lopez"
                },
                {
                    "id": 1,
                    "name": "Marks Herman"
                },
                {
                    "id": 2,
                    "name": "Salas Meyers"
                }
            ],
            "greeting": "Hello, Katelyn Bruce! You have 6 unread messages.",
            "favoriteFruit": "banana"
        },
        {
            "_id": "5c77256b46d29e2d9705b2d7",
            "index": 4,
            "guid": "d71fd89c-7191-4186-b653-ccffddbf1d04",
            "isActive": true,
            "balance": "$3,055.55",
            "picture": "http://placehold.it/32x32",
            "age": 25,
            "eyeColor": "blue",
            "name": "Hodges Espinoza",
            "gender": "male",
            "company": "MEMORA",
            "email": "hodgesespinoza@memora.com",
            "phone": "+1 (973) 425-2889",
            "address": "199 Florence Avenue, Alamo, Vermont, 1187",
            "about": "Lorem culpa sunt esse est ut et ea ea cillum enim commodo. Quis laborum officia deserunt amet magna culpa irure. Pariatur aliquip sint commodo esse in pariatur sint ullamco adipisicing reprehenderit id ea. Occaecat officia veniam nostrud sit ex velit incididunt. Tempor do aliquip cillum irure adipisicing. Ut ut sit voluptate voluptate adipisicing labore.\r\n",
            "registered": "2014-01-22T04:07:00 +08:00",
            "latitude": -10.760669,
            "longitude": 10.497581,
            "tags": [
                "laboris",
                "ut",
                "qui",
                "magna",
                "minim",
                "dolore",
                "ullamco"
            ],
            "friends": [
                {
                    "id": 0,
                    "name": "Pacheco Matthews"
                },
                {
                    "id": 1,
                    "name": "Lori Battle"
                },
                {
                    "id": 2,
                    "name": "Anthony Cote"
                }
            ],
            "greeting": "Hello, Hodges Espinoza! You have 5 unread messages.",
            "favoriteFruit": "banana"
        },
        {
            "_id": "5c77256b9fe07cb9744a80c3",
            "index": 5,
            "guid": "38d6f176-ef52-4aa1-9747-613a43212bc9",
            "isActive": true,
            "balance": "$3,356.10",
            "picture": "http://placehold.it/32x32",
            "age": 36,
            "eyeColor": "blue",
            "name": "Booth Beard",
            "gender": "male",
            "company": "PHARMEX",
            "email": "boothbeard@pharmex.com",
            "phone": "+1 (947) 526-2640",
            "address": "294 Marconi Place, Kanauga, Idaho, 8886",
            "about": "Elit ad aliquip ad nostrud eu culpa elit voluptate consectetur qui excepteur. Fugiat aute voluptate minim exercitation sit. Dolore deserunt pariatur sit nulla. Duis velit cillum ipsum anim velit adipisicing nulla nulla sit magna amet. Aliqua nisi quis sit consectetur aliquip dolor nisi excepteur deserunt occaecat reprehenderit quis nulla. Eiusmod fugiat culpa ullamco tempor cillum commodo. Veniam quis velit aliqua do ea velit enim velit ea irure aliquip ex laboris.\r\n",
            "registered": "2017-05-28T05:18:10 +07:00",
            "latitude": -62.281245,
            "longitude": 169.880199,
            "tags": [
                "deserunt",
                "amet",
                "eu",
                "pariatur",
                "consequat",
                "non",
                "exercitation"
            ],
            "friends": [
                {
                    "id": 0,
                    "name": "Lorna Barker"
                },
                {
                    "id": 1,
                    "name": "Alyssa Carroll"
                },
                {
                    "id": 2,
                    "name": "Marcy Jacobs"
                }
            ],
            "greeting": "Hello, Booth Beard! You have 6 unread messages.",
            "favoriteFruit": "strawberry"
        }
    ];

    t.deepEqual( parseExpression( encode( largeDataPayload ) ), largeDataPayload );

    t.end();
} );


clojure/src/sexp_parser/core.clj
(ns sexp-parser.core
  (:require [clojure.string :as str]
            [clojure.edn :as edn]))

(defn tokenize [expression-string]
  (edn/read-string expression-string))

(defn map-or-self [item]
  (try
    (let [as-map (apply hash-map item)]
      (if (and (every? #(= clojure.lang.Keyword (type %1)) (keys as-map))
               (< 0 (count as-map)))
        as-map
        item))
    (catch Exception e item)))

(defn make-maps [item]
  "Map over elements of the array, turn them in to hash maps if possible."
  (if (coll? item)
    (map-or-self (map make-maps item))
    item))

(defn parse [data]
  (make-maps (tokenize data)))

(defn encode-atom [data]
  (cond
    (= false data)
    "NIL"
    (= true data)
    "T"
    :else
    (pr-str data)))

(declare encode)
(defn encode-array [data]
  (str "("
       (str/join " " (map encode data))
       ")"))

(defn encode-map [data]
  (pr-str data))

(defn encode [data]
  (->
   (cond
     (or (vector? data) (list? data))
     (encode-array data)
     (map? data)
     (encode-map data)
     :else
     (encode-atom data))
   (str/replace #"," "")
   (str/replace #"[\[\{]" "(")
   (str/replace #"[\]\}]" ")")
   (str/replace #"false" "NIL")
   (str/replace #"true" "T")))


javascript/benchmarks.js
const largeDataPayload = [
    {
        "_id": "5c77256b120e71539a4543a7",
        "index": 0,
        "guid": "c9815eb2-fdc5-42f8-bbb3-c2dadaf4c1b8",
        "isActive": true,
        "balance": "$1,556.23",
        "picture": "http://placehold.it/32x32",
        "age": 40,
        "eyeColor": "brown",
        "name": "Dora Weiss",
        "gender": "female",
        "company": "CIRCUM",
        "email": "doraweiss@circum.com",
        "phone": "+1 (957) 431-3727",
        "address": "219 Chase Court, Hebron, New Jersey, 6091",
        "about": "Quis ipsum eu deserunt ut mollit est eu duis commodo reprehenderit cupidatat nulla ad aliquip. Tempor nulla in consequat est voluptate veniam excepteur pariatur sunt sint. Veniam laborum aliqua esse elit qui est exercitation velit. Mollit qui exercitation nulla elit ex ad aute laborum reprehenderit. Quis ipsum proident velit do. Dolor dolore incididunt aliquip culpa cillum enim. Id incididunt ullamco aliquip id mollit elit ullamco qui duis ut mollit.\r\n",
        "registered": "2014-08-01T12:47:41 +07:00",
        "latitude": -82.633003,
        "longitude": 4.637068,
        "tags": [
            "fugiat",
            "eu",
            "Lorem",
            "laborum",
            "fugiat",
            "pariatur",
            "ex"
        ],
        "friends": [
            {
                "id": 0,
                "name": "Lynette Mccarthy"
            },
            {
                "id": 1,
                "name": "Leila Medina"
            },
            {
                "id": 2,
                "name": "Kasey Moss"
            }
        ],
        "greeting": "Hello, Dora Weiss! You have 9 unread messages.",
        "favoriteFruit": "banana"
    },
    {
        "_id": "5c77256ba6e9551c135fc72f",
        "index": 1,
        "guid": "ec2b621f-e35a-4447-9ecb-c3bee2feeeea",
        "isActive": true,
        "balance": "$1,782.51",
        "picture": "http://placehold.it/32x32",
        "age": 30,
        "eyeColor": "blue",
        "name": "Karen Hernandez",
        "gender": "female",
        "company": "CONCILITY",
        "email": "karenhernandez@concility.com",
        "phone": "+1 (853) 530-3600",
        "address": "478 Schenck Street, Roland, Federated States Of Micronesia, 8478",
        "about": "Esse id do irure ipsum quis occaecat cillum pariatur est ipsum esse exercitation reprehenderit aliqua. Ea nostrud fugiat sunt sunt. Ut aute voluptate consequat occaecat cupidatat magna irure. Occaecat sunt fugiat laboris est labore laborum dolor reprehenderit irure ipsum officia voluptate id nisi. Duis consequat minim nostrud magna nostrud consequat eiusmod nulla qui sunt incididunt cupidatat. Lorem ullamco aliqua officia adipisicing eiusmod labore incididunt irure aute nulla aliquip laborum cillum.\r\n",
        "registered": "2016-03-14T11:34:25 +07:00",
        "latitude": -5.325438,
        "longitude": 18.0786,
        "tags": [
            "exercitation",
            "sit",
            "ad",
            "sit",
            "fugiat",
            "enim",
            "non"
        ],
        "friends": [
            {
                "id": 0,
                "name": "Marilyn Mooney"
            },
            {
                "id": 1,
                "name": "Cindy Nunez"
            },
            {
                "id": 2,
                "name": "Ramirez Gilmore"
            }
        ],
        "greeting": "Hello, Karen Hernandez! You have 4 unread messages.",
        "favoriteFruit": "strawberry"
    },
    {
        "_id": "5c77256bef0f2c88b63e1c2a",
        "index": 2,
        "guid": "7f91e154-37c3-40ca-b240-71854027eb7f",
        "isActive": true,
        "balance": "$2,210.88",
        "picture": "http://placehold.it/32x32",
        "age": 26,
        "eyeColor": "brown",
        "name": "Short Page",
        "gender": "male",
        "company": "PLEXIA",
        "email": "shortpage@plexia.com",
        "phone": "+1 (834) 421-2685",
        "address": "129 Taaffe Place, Belmont, New Mexico, 7156",
        "about": "Ipsum minim eiusmod cupidatat irure. Adipisicing voluptate ex mollit reprehenderit. Occaecat aliqua proident enim cillum adipisicing. Do adipisicing incididunt sit non cillum. Aute duis dolor incididunt nulla sit duis cillum non id.\r\n",
        "registered": "2019-01-20T12:58:27 +08:00",
        "latitude": -10.195242,
        "longitude": 7.95824,
        "tags": [
            "occaecat",
            "proident",
            "pariatur",
            "officia",
            "eiusmod",
            "velit",
            "laboris"
        ],
        "friends": [
            {
                "id": 0,
                "name": "Naomi Jennings"
            },
            {
                "id": 1,
                "name": "Blankenship Pruitt"
            },
            {
                "id": 2,
                "name": "Collins Carney"
            }
        ],
        "greeting": "Hello, Short Page! You have 3 unread messages.",
        "favoriteFruit": "strawberry"
    },
    {
        "_id": "5c77256b4b4fbbd1e17ea23e",
        "index": 3,
        "guid": "39955cf0-3a93-4fa8-a85f-882bdcfc43c7",
        "isActive": true,
        "balance": "$3,613.59",
        "picture": "http://placehold.it/32x32",
        "age": 30,
        "eyeColor": "green",
        "name": "Katelyn Bruce",
        "gender": "female",
        "company": "ZOINAGE",
        "email": "katelynbruce@zoinage.com",
        "phone": "+1 (962) 514-3834",
        "address": "201 Irwin Street, Succasunna, Texas, 1045",
        "about": "Commodo elit laborum amet mollit. Non aute irure esse esse est anim ea mollit qui commodo. Do dolor occaecat Lorem incididunt qui anim commodo sunt. Occaecat cupidatat occaecat eiusmod dolore officia excepteur culpa laborum id eiusmod velit. Voluptate non cillum id nulla deserunt consectetur esse sunt quis. Eiusmod et anim occaecat sunt aute do aliqua id reprehenderit voluptate occaecat duis aliqua. Veniam nostrud velit laboris excepteur occaecat est elit.\r\n",
        "registered": "2014-08-27T03:12:23 +07:00",
        "latitude": 61.029178,
        "longitude": 51.670902,
        "tags": [
            "consectetur",
            "deserunt",
            "irure",
            "occaecat",
            "voluptate",
            "sint",
            "esse"
        ],
        "friends": [
            {
                "id": 0,
                "name": "Daniel Lopez"
            },
            {
                "id": 1,
                "name": "Marks Herman"
            },
            {
                "id": 2,
                "name": "Salas Meyers"
            }
        ],
        "greeting": "Hello, Katelyn Bruce! You have 6 unread messages.",
        "favoriteFruit": "banana"
    },
    {
        "_id": "5c77256b46d29e2d9705b2d7",
        "index": 4,
        "guid": "d71fd89c-7191-4186-b653-ccffddbf1d04",
        "isActive": true,
        "balance": "$3,055.55",
        "picture": "http://placehold.it/32x32",
        "age": 25,
        "eyeColor": "blue",
        "name": "Hodges Espinoza",
        "gender": "male",
        "company": "MEMORA",
        "email": "hodgesespinoza@memora.com",
        "phone": "+1 (973) 425-2889",
        "address": "199 Florence Avenue, Alamo, Vermont, 1187",
        "about": "Lorem culpa sunt esse est ut et ea ea cillum enim commodo. Quis laborum officia deserunt amet magna culpa irure. Pariatur aliquip sint commodo esse in pariatur sint ullamco adipisicing reprehenderit id ea. Occaecat officia veniam nostrud sit ex velit incididunt. Tempor do aliquip cillum irure adipisicing. Ut ut sit voluptate voluptate adipisicing labore.\r\n",
        "registered": "2014-01-22T04:07:00 +08:00",
        "latitude": -10.760669,
        "longitude": 10.497581,
        "tags": [
            "laboris",
            "ut",
            "qui",
            "magna",
            "minim",
            "dolore",
            "ullamco"
        ],
        "friends": [
            {
                "id": 0,
                "name": "Pacheco Matthews"
            },
            {
                "id": 1,
                "name": "Lori Battle"
            },
            {
                "id": 2,
                "name": "Anthony Cote"
            }
        ],
        "greeting": "Hello, Hodges Espinoza! You have 5 unread messages.",
        "favoriteFruit": "banana"
    },
    {
        "_id": "5c77256b9fe07cb9744a80c3",
        "index": 5,
        "guid": "38d6f176-ef52-4aa1-9747-613a43212bc9",
        "isActive": true,
        "balance": "$3,356.10",
        "picture": "http://placehold.it/32x32",
        "age": 36,
        "eyeColor": "blue",
        "name": "Booth Beard",
        "gender": "male",
        "company": "PHARMEX",
        "email": "boothbeard@pharmex.com",
        "phone": "+1 (947) 526-2640",
        "address": "294 Marconi Place, Kanauga, Idaho, 8886",
        "about": "Elit ad aliquip ad nostrud eu culpa elit voluptate consectetur qui excepteur. Fugiat aute voluptate minim exercitation sit. Dolore deserunt pariatur sit nulla. Duis velit cillum ipsum anim velit adipisicing nulla nulla sit magna amet. Aliqua nisi quis sit consectetur aliquip dolor nisi excepteur deserunt occaecat reprehenderit quis nulla. Eiusmod fugiat culpa ullamco tempor cillum commodo. Veniam quis velit aliqua do ea velit enim velit ea irure aliquip ex laboris.\r\n",
        "registered": "2017-05-28T05:18:10 +07:00",
        "latitude": -62.281245,
        "longitude": 169.880199,
        "tags": [
            "deserunt",
            "amet",
            "eu",
            "pariatur",
            "consequat",
            "non",
            "exercitation"
        ],
        "friends": [
            {
                "id": 0,
                "name": "Lorna Barker"
            },
            {
                "id": 1,
                "name": "Alyssa Carroll"
            },
            {
                "id": 2,
                "name": "Marcy Jacobs"
            }
        ],
        "greeting": "Hello, Booth Beard! You have 6 unread messages.",
        "favoriteFruit": "strawberry"
    }
];

const getTime = () => {
    return process.hrtime()[1];
};

const sexp = require('./index.js');

let iterations = 1000;

let encodedJson = JSON.stringify( largeDataPayload );

let jsonEncodeTime = 0;
for ( let i = 0; i < iterations; i++ ) {
    const jsonEncodeStart = getTime();
    encodedJson = JSON.stringify( largeDataPayload );
    jsonEncodeTime += ( getTime() - jsonEncodeStart ) / iterations;
}
// jsonEncodeTime /= iterations;

let parsedJson = JSON.parse( encodedJson );

let jsonParseTime = 0;
for ( let i = 0; i < iterations; i++ ) {
    const jsonParseStart = getTime();
    parsedJson = JSON.parse( encodedJson );
    jsonParseTime += ( getTime() - jsonParseStart ) / iterations;
}
// jsonParseTime /= iterations;

let encodedSexp = sexp.encode( largeDataPayload );

let sexpEncodeTime = 0;
for ( let i = 0; i < iterations; i++ ) {
    const sexpEncodeStart = getTime();
    encodedSexp = sexp.encode( largeDataPayload );
    sexpEncodeTime += ( getTime() - sexpEncodeStart ) / iterations;
}
// sexpEncodeTime /= iterations;

let parsedSexp = sexp.parse( encodedSexp );

let sexpParseTime = 0;
for ( let i = 0; i < iterations; i++ ) {
    const sexpParseStart = getTime();
    parsedSexp = sexp.parse( encodedSexp );
    sexpParseTime += ( getTime() - sexpParseStart ) / iterations;
}
// sexpParseTime /= iterations;

console.log( `JSON parse:  ${ jsonParseTime }` );
console.log( `JSON encode: ${ jsonEncodeTime }` );
console.log( `sexp parse:  ${ sexpParseTime }` );
console.log( `sexp encode: ${ sexpEncodeTime }` );
console.log( `parse is ${ sexpParseTime / jsonParseTime } slower` );
console.log( `encode is ${ sexpEncodeTime / jsonEncodeTime } slower` );
